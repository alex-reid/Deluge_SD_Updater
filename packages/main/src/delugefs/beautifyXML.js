/* MODIFIED FROM

  name: xml-beautifier,
  version: 0.5.0,
  title: XML Beautifier,
  description: Beautifies XML documents by putting each tag and text node on their own line and correctly indents everything,

  url: https://github.com/svenheden/xml-beautifier.git
  author: Jonathan Svenheden <jonathan@svenheden.com>,
  license: MIT,
*/

const repeat = require('repeat-string');

const splitOnTags = function splitOnTags(str) {
  return str.split(/(<\/?[^>]+>)/g).filter(function (line) {
    return line.trim() !== '';
  });
};
const isTag = function isTag(str) {
  return /<[^>!]+>/.test(str);
};
const isXMLDeclaration = function isXMLDeclaration(str) {
  return /<\?[^?>]+\?>/.test(str);
};
const isClosingTag = function isClosingTag(str) {
  return /<\/+[^>]+>/.test(str);
};
const isSelfClosingTag = function isSelfClosingTag(str) {
  return /<[^>]+\/>/.test(str);
};
const isOpeningTag = function isOpeningTag(str) {
  return isTag(str) && !isClosingTag(str) && !isSelfClosingTag(str) && !isXMLDeclaration(str);
};

function getAttrs(input) {
  const regex = /<(\w+)([^>]*?)(\/?)>/;
  const match = input.match(regex);

  if (match) {
    const openingTag = `<${match[1]}`;
    const attributesString = match[2].trim();
    const hasClosingSlash = !!match[3];
    const closingTag = hasClosingSlash ? '/>' : '>';

    // Extracting individual attributes as an array of strings
    const attributeRegex = /(\w+="[^"]*")/g;
    const attributes = attributesString.match(attributeRegex) || [];
    if (attributes.length > 2) {
      return {openingTag, attributes, closingTag};
    }
  }
  return null;
}

const tagsToSkip = ['modeNote', 'selectedDrumIndex'];

function skipTags(item, endpos = false) {
  const tagName = item.match(/<(\/?)(\w+)/);
  if (tagName?.[2]) {
    if (tagName[1] == '/' && !endpos) return false;
    if (tagName[1] == '' && endpos) return false;
    return tagsToSkip.includes(tagName[2]);
  }
  return false;
}

export default function beautifyXML(xml, indent = '    ') {
  let depth = 0;
  let ignoreMode = false;
  let deferred = [];

  return splitOnTags(xml)
    .map(function (item) {
      if (skipTags(item)) {
        item = repeat(indent, depth) + item;
        ignoreMode = true;
      }
      if (skipTags(item, true)) {
        ignoreMode = false;
        deferred.push(item);
        const cdataBlock = deferred.join('');
        deferred = [];
        return cdataBlock;
      }
      if (ignoreMode) {
        deferred.push(item);
        return null;
      }

      const attrs = getAttrs(item);
      let line;

      // removes any pre-existing whitespace chars at the end or beginning of the item
      item = item.replace(/^\s+|\s+$/g, '');

      if (isClosingTag(item)) {
        depth--;
      }

      if (attrs) {
        line = repeat(indent, depth) + attrs.openingTag + '\n' + repeat(indent, depth + 1);
        line += attrs.attributes.join('\n' + repeat(indent, depth + 1));
        line += attrs.closingTag;
      } else {
        line = repeat(indent, depth) + item;
      }

      if (isOpeningTag(item)) {
        depth++;
      }

      return line;
    })
    .filter(function (c) {
      return c;
    })
    .join('\n');
}
