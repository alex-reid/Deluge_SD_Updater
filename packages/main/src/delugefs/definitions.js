const letterToIndexMap = {
  a: 0,
  b: 1,
  c: 2,
  d: 3,
  e: 4,
  f: 5,
  g: 6,
  h: 7,
  i: 8,
  j: 9,
  k: 10,
  l: 11,
  m: 12,
  n: 13,
  o: 14,
  p: 15,
  q: 16,
  r: 17,
  s: 18,
  t: 19,
  u: 20,
  v: 21,
  w: 22,
  x: 23,
  y: 24,
  z: 25,
};

const indexToLetterMap = [
  'a',
  'b',
  'c',
  'd',
  'e',
  'f',
  'g',
  'h',
  'i',
  'j',
  'k',
  'l',
  'm',
  'n',
  'o',
  'p',
  'q',
  'r',
  's',
  't',
  'u',
  'v',
  'w',
  'x',
  'y',
  'z',
];

const newNames = {
  SYNTHS: [
    '000 Rich Saw Bass',
    '001 Sync Bass',
    '002 Basic Square Bass',
    '003 Synthwave Bass',
    '004 Dubby Bass',
    '005 Sweet Mono Bass',
    '006 Vaporwave Bass',
    '007 Detuned Saw Bass',
    '008 FM Rich Distorted Bass',
    '009 Hoover Bass',
    '010 Gravel Basscamp',
    '011 Dubstep Bass',
    '012 Blunt Sync Bass',
    '013 Trap Bass 1',
    '014 Trap Bass 2',
    '015 Resonant Filter Bass',
    '016 Dark Saturated Bass',
    '017 Impact Saw Lead',
    '018 Rich Saw Lead',
    '019 Fizzy Strings',
    '020 Soft Saw Lead',
    '021 80s TV Lead',
    '022 Rich Filter LFO Lead',
    '023 Analog Mono Wow',
    '024 Warble Bass Pluck',
    '025 Soft Synth Organ',
    '026 PW Organ',
    '027 PW Envelope',
    '028 PWM',
    '029 Chiptune Trill',
    '030 Distant Porta',
    '031 Nasal Choir',
    '032 Bandpass Choir',
    '033 Rich Square',
    '034 Square Choir',
    '035 Bell Lead & Bass',
    '036 Analog Ambient Square',
    '037 Echo Chord',
    '038 Vapor Arp',
    '039 Detuned Retriggering Saws',
    '040 Spacer Leader',
    '041 Zithar - Vibed',
    '042 High Triangle',
    '043 Square Porta',
    '044 8-Bit Lead',
    '045 Square Sync',
    '046 Saw Sync',
    '047 Basic Dirty Bass',
    '048 Thin Pulse Bass',
    '049 Basic FM',
    '050 FM Basic Bass',
    '051 FM Rich Brass',
    '052 Soft FM',
    '053 Detuned FM Horns',
    '054 Ghostly Sines',
    '055 FM Theremin',
    '056 FM Bell Modulation',
    '057 FM Lead',
    '058 FM Rising Attack',
    '059 Distorted Lead Guitar',
    '060 Bass Guitar',
    '061 Blown-Staccato-Panpipes',
    '062 Trumpet',
    '063 Tuba',
    '064 Reeds-Flute-Oboe',
    '065 Cello',
    '066 Violin',
    '067 Marimba',
    '068 FM Bells 1',
    '069 FM Bells 2',
    '070 Glockenspiel',
    '071 Rhodes',
    '072 Kyoto Phono',
    '073 Piano',
    '074 Electric Piano',
    '075 Electric Piano With Strings',
    '076 Organ',
    '077 FM Perc-Organ',
    '078 House',
    '079 Phased Arper',
    '080 House',
    '081 Xylophone Big Bass',
    '082 Short Sharp Delay',
    '083 Dark Chorus',
    '084 FM Narrow Band',
    '085 Deep Fizz',
    '086 Techno Organ',
    '087 Define Leader',
    '088 Yelp Chords',
    '089 Degraded Retro Lead',
    '090 FM Organ',
    '091 FM Ricochet',
    '092 Degraded Tremolo',
    '093 FM Distorted Bells',
    '094 Ambient Occlusion Lead',
    '095 Harsh FM Feedback',
    '096 FM Guitar Power Chord',
    '097 Saturated Filter',
    '098 Saturated Sync',
    '099 Overdrive Reese Sync',
    '100 Noise Lead',
    '101 Atebit',
    '102 Harsh 5th',
    '103 Sci-fi Chaos',
    '104 Alien Vomit',
    '105 Attack Bass',
    '106 Hang Drum',
    '107 FM LPG Percussion',
    '108 Robo Arp',
    '109 Talking Arp',
    '110 Crystalline Ringmod',
    '111 Satellite Drum',
    '112 Hard Tech Beat',
    '113 Bio Lab',
    '114 Sootheerio',
    '115 Sounds After Take-off',
    '116 Evolving Frequencies',
    '117 Belledy',
    '118 Small Bridge Pad',
    '119 Stars Of The Bin Pad',
    '120 High Harsh Pad',
    '121 Tiny Lights',
    '122 Majestic Synth Orchestra',
    '123 Space Dust',
    '124 Filter Modulation Pad',
    '125 Evolving Pad',
    '126 Dark FM Pad',
    '127 Alien Larvae',
    '128 Lunar Landing',
    '129 Sci-fi Scenic',
    '130 Dark Strings',
    '131 Warm Strings',
    '132 Organ Strings',
    '133 80s Strings',
    '134 Melody String',
    '135 Soothing Growth Pad',
    '136 Synthwave Pad',
    '137 Epic Saw Modulation Pad',
    '138 Brassy Pad',
    '139 Detuned Saw Pad',
    '140 Slow Aural Swells',
    '141 Ringmod Pad',
    '142 Phaser',
    '143 Chillout Pad',
    '144 Sweep Chords',
    '145 Eerie High Pad',
    '146 Atmospheric Squares Pad',
    '147 Resonant Filter Pad',
    '148 Warm 5th Pad',
    '149 Cold 5th Pad',
    '150 Vaporwave Pad',
    '151 Radiant FM Pad',
    '152 Small Jet Pad',
    '153 FM Modulation Pad',
    '154 Rich FM Pad 1',
    '155 Rich FM Pad 2',
    '156 Rich FM Pad 3',
    '157 Rich FM Pad 4',
    '158 Tempo-Synced LFO',
    '159 80s Bass Rhythm',
    '160 Synthwave Bass Arp',
    '161 Synthwave Vibrato Arp',
    '162 Busy Arp',
    '163 Crisp Pop Arp',
    '164 Study Arp',
    '165 Acid Arp',
    '166 Harpsichord Cyborg',
    '167 FM Metallic Bass Arp',
    '168 Hang Drum',
    '169 Double Bass',
    '170 Sitar',
  ],
  KITS: [
    '000 TR-808',
    '001 DDD-1',
    '002 SDS-5',
    '003 TR-909',
    '004 R-50',
    '005 R-100',
    '006 LD',
    '007 HR-16B',
    '008 SCDT',
    '009 RX-5',
    '010 XV-5080',
    '011 KR-55',
    '012 HR-II',
    '013 AT Rhythm',
    '014 CR-78',
    '015 Andrew Stirton Frugal',
    '016 Electronisounds 1',
    '017 Electronisounds 2',
    '018 Electronisounds 3',
    '019 Fairburg',
    '020 Leonard Ludvigsen Beatbox',
    '021 hodeur 1',
    '022 hodeur 2',
    '023 hodeur 3',
    '024 James R Closs 1',
    '025 James R Closs 2',
    '026 Amiga909',
    '027 Reciprocal Sound',
    '028 Danny Taurus',
    '029 Danny Taurus 2',
    '030 Chaz Bundick',
    '031 Reuben Winter',
    '032 Kody Nielson',
    '033 Alfred Darlington',
    '034 Travis Egedy',
    '035 Sjionel Timu',
    '036 Stefanie Franciotti',
    '037 Stephanie Engelbrecht',
    '038 Jonathan Snipes (FX)',
    '039 Campbell Kneale',
    '040 John Atkinson',
    '041 Jonathan Snipes (Waterfalls)',
    '042 Phil Elverum',
  ],
};

const typeMappings = [
  {
    type: 'synths',
    folder: 'SYNTHS',
    xml: 'sound',
    file: 'SYNT',
    pretty: 'Synth',
  },
  {type: 'kits', folder: 'KITS', xml: 'kit', file: 'KIT', pretty: 'Kit'},
  {type: 'songs', folder: 'SONGS', xml: 'song', file: 'SONG', pretty: 'Song'},
];

export const FORMATS = {
  /** Old style slot + sub slot. */
  OLD: 'old style slot subslot',
  /** V4-ish. Just three numbers and folder  */
  NUMBERS_ONLY: 'three numbers only',
  /** A new name with an added suffix */
  NEW_SUFFIX: 'new style with suffix',
  /** V4 file and folder */
  NEW: 'new v4',
  /** just a preset name */
  JUST_NAME: 'name only',
  /** Hopefully never see this one */
  UNKNOWN: 'unknown type',
};

export {letterToIndexMap, indexToLetterMap, newNames, typeMappings};
