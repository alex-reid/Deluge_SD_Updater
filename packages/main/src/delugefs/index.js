const fileSystem = require('./fileSystemClass');

const D = new fileSystem(
  // "/Volumes/DELUGE"
  //"./Deluge_v2",
  //"./Deluge_v4",
  './DelugeSD',
  // "/Users/alexreid/work/deluge-node/Deluge+OLED+V4p1p0+factory+card+contents",
  {
    renameToV4: true,
    prettyNames: false,
  },
);

D.init()
  .then(() => {
    D.rewriteInstrumentsToV4()
      .then(() => console.log('Sucessfully renamed Synths and Kits'))
      .catch(err => console.log('Failed remaning Synths and Kits', err));
    D.rewriteSongsToV4()
      .then(() => console.log('Sucessfully rewrote Songs'))
      .catch(err => console.log('Failed Rewriting songs', err));
  })
  .catch(err => console.error(err));
