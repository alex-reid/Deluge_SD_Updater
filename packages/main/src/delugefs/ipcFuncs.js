import {prettyName} from './utils';

function getMainDelugeInfo(D) {
  return {
    initialised: true,
    kits: D.files.kits.map(kit => ({
      path: kit.path,
      rootPath: kit.rootPath,
      rewriteName: '',
      oldName: kit.fileName,
      newName: kit.newName,
      prettyName: prettyName(kit.presetName) || '',
      presetType: kit.presetType,
      presetName: kit.presetName,
      songIDs: Array.from(kit.songIDs),
    })),
    synths: D.files.synths.map(synth => ({
      path: synth.path,
      rootPath: synth.rootPath,
      rewriteName: '',
      oldName: synth.fileName,
      newName: synth.newName,
      prettyName: prettyName(synth.presetName) || '',
      presetType: synth.presetType,
      presetName: synth.presetName,
      songIDs: Array.from(synth.songIDs),
    })),
    songs: D.files.songs.map(song => {
      return {
        name: song.fileName,
        path: song.path,
        clipsLength: song.clips.length,
        firmwareVersion: song.firmwareVersion,
        shouldUpdate: true,
        previewData: song.previewData,
        instruments: song.instruments.reduce(
          (acc, curr) => [
            ...acc,
            {
              type: curr.presetType,
              presetName: curr.presetName,
              presetFolder: curr.presetFolder,
              presetSlot: curr.presetSlot,
              presetSubSlot: curr.presetSubSlot,
              formatType: curr.formatType,
              patchName: curr.patchName,
              patchSuffix: curr.patchSuffixClean,
              patchSuffixOld: curr.patchSuffix,
              usedInClips: song.getInstrumentClipIndexs(curr),
              rewriteName: curr.rewriteName,
              rewriteFolder: curr.rewriteFolder,
              prettyName: prettyName(curr.patchName) + curr.patchSuffixClean || '',
              soundID: curr.soundID,
              isNewSound: curr.soundID == 'new',
            },
          ],
          [],
        ),
      };
    }),
  };
}

const sendMainDelugeInfo = (D, browserWindow) =>
  browserWindow.webContents.send('files', getMainDelugeInfo(D));

const sendErrorMain = (error, browserWindow) => browserWindow.webContents.send('send-error', error);

export {getMainDelugeInfo, sendMainDelugeInfo, sendErrorMain};
