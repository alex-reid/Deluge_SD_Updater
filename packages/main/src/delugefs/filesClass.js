import {load} from 'cheerio';
import fs from 'fs/promises';
import path from 'path';
import {getOldTypeAndNumber, getFolderFromFileType, getNameAndSuffix, getPath} from './utils';

import {Instrument, Clip} from './nodesClass';
import {newNames} from './definitions';

/**
 * Class representing an XML file on the deluge
 */
class File {
  constructor(filePath, fileName, rootPath) {
    this.path = getPath(rootPath, filePath); //filePath.replace(rootPath + path.sep, '');
    this.rootPath = rootPath;
    this.fullFileName = fileName;
    this.fileName = fileName.replace(/\.xml$/i, '');
    this.XML = null;
    this.isLoaded = false;
    console.log(getPath(rootPath, filePath));
  }

  async loadXML() {
    try {
      const xmlContent = await fs.readFile(
        path.join(this.rootPath, this.path, this.fullFileName),
        'utf-8',
      );
      this.XML = load(
        xmlContent,
        {
          xmlMode: true, // Set xmlMode to true for XML parsing
          pretty: true, // Enable pretty printing
          onParseError: err => console.error('parse error', err),
        },
        false,
      );
      this.isLoaded = true;
      this.onXMLLoaded();
    } catch (error) {
      console.error(
        'Error loading or parsing XML file:',
        path.join(this.rootPath, this.path, this.fileName),
        '\n',
        error,
      );
    }
  }

  onXMLLoaded() {}

  async saveXML() {
    // Save the modified XML to a new file
    const modifiedXml = this.XML.xml();
    const newXmlFilePath = path.join(this.rootPath, this.path, this.fileName + '_v4.XML');
    fs.writeFile(newXmlFilePath, modifiedXml, 'utf-8');
  }
}

/**
 * Class representing a song XML file on the deluge
 * @extends File
 */
class Song extends File {
  constructor(path, fileName, mappings, rootPath) {
    super(path, fileName, rootPath);
    this.mappings = mappings;
    this.elementsWithAttributes = null;
    this.elementsWithNames = null;
    /** @type {Instrument[]} */
    this.instruments = [];
    /** @type {Clip[]} */
    this.clips = [];
    this.firmwareVersion = null;
  }

  onXMLLoaded() {
    this.getFirmwareVersion();
    this.getInstruments();
    this.getClips();
  }

  getFirmwareVersion() {
    if (this.isLoaded) {
      this.firmwareVersion = this.XML('[firmwareVersion]').attr('firmwareVersion')?.split('.') || [
        0, 0, 0,
      ];
    }
  }

  getInstruments() {
    if (this.isLoaded) {
      this.instruments = [];
      this.XML('instruments > sound, instruments > kit').each((_i, instrument) => {
        this.instruments.push(new Instrument(instrument, this.XML));
      });
    }
    return this.instruments.length;
  }

  validate() {
    let valid = true;
    this.addClipIdsToInstruments();
    let instClips = 0;
    this.clips.forEach(clip => {
      if (clip.presetType != 'notsound') instClips++;
    });
    let clipsFound = 0;
    this.instruments.forEach(i => {
      if (i.clips.length < 1) {
        valid = false;
        throw new Error('found an orphaned sound in ' + this.fileName);
      }
      clipsFound += i.clips.length;
    });
    if (instClips != clipsFound) {
      valid = false;
      throw new Error('found an orphaned clip in ' + this.fileName);
    }
    return valid;
  }

  rewriteInstrumentsAndClipsXMLAttributes(newNames) {
    if (this.instruments.length !== newNames.length) {
      throw new Error('new names are not the same length as instruments');
    }
    this.instruments.forEach((instrument, index) => {
      const {rewriteName, rewriteFolder} = newNames[index];
      if (!rewriteName || !rewriteFolder) {
        throw new Error('invalid data in new names array');
      }
      instrument.rewritePresetToV4(rewriteName, rewriteFolder);
      instrument.clips.forEach(clip => {
        this.clips[clip].rewritePresetToV4(rewriteName, rewriteFolder);
      });
    });
  }

  getClips() {
    if (this.isLoaded) {
      this.clips = [];
      this.XML('sessionClips instrumentClip').each((_i, clip) => {
        this.clips.push(new Clip(clip, this.XML));
      });
    }
    return this.clips.length;
  }

  getInstrumentClipIndexs(instrument) {
    return this.clips.reduce((acc, clip, index) => {
      if (
        instrument.presetType == clip.presetType &&
        instrument.presetSlot == clip.presetSlot &&
        instrument.presetSubSlot == clip.presetSubSlot &&
        instrument.presetName == clip.presetName
      ) {
        return [...acc, index];
      }
      return acc;
    }, []);
  }

  addClipIdsToInstruments() {
    for (const instrument of this.instruments) {
      instrument.clips = this.getInstrumentClipIndexs(instrument);
    }
  }
}

/**
 * Class representing an Instrument XML file on the deluge
 * @extends File
 */
class Sound extends File {
  constructor(path, fileName, rootPath) {
    super(path, fileName, rootPath);
    /** @type {number|null} */
    this.soundID = null;
    this.presetName = fileName.replace(/\.xml$/i, '');
    this.newName = this.getNewName();
    this.songIDs = new Set();
    /** @type {('KIT'|'SYNT')} */
    this.presetType = '';
  }

  getNewName() {
    const [name, , suffix] = getNameAndSuffix(this.fileName);
    if (name) {
      let newName = name;
      const fileData = getOldTypeAndNumber(name);
      if (fileData)
        newName = newNames[getFolderFromFileType(fileData.presetType)][fileData.presetSlot] || name;
      return newName + suffix;
    }
    return '';
  }

  renameFileToV4() {
    const newName = this.getNewName();
    if (this.isOldName) {
      fs.rename(
        path.join(this.rootPath, this.path, this.fileName + '.XML'),
        path.join(this.rootPath, this.path, newName + '.XML'),
      )
        .then(() => console.log('Renamed', this.fileName + '.XML', 'to', newName + '.XML'))
        .catch(err => console.error(err));
    }
  }
}
/**
 * Class representing an Kit XML file on the deluge
 * @extends Sound
 */
class Kit extends Sound {
  constructor(path, fileName, rootPath) {
    super(path, fileName, rootPath);
    this.presetType = 'KIT';
  }
}
/**
 * Class representing an Synth XML file on the deluge
 * @extends Sound
 */
class Synth extends Sound {
  constructor(path, fileName, rootPath) {
    super(path, fileName, rootPath);
    this.presetType = 'SYNT';
  }
}

export {File, Song, Sound, Kit, Synth};
