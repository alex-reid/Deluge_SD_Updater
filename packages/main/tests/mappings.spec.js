import {afterAll, beforeAll, expect, test} from 'vitest';
import fileSystem from '../src/delugefs/fileSystemClass';
import path from 'path';
import {getMainDelugeInfo} from '../src/delugefs/ipcFuncs';

import {fakeFS} from '../../../fake_filesystem/setupDummyFiles';

let Deluge, dummy_dir;
const testfolder = path.resolve(__dirname);

expect.addSnapshotSerializer({
  print(val) {
    // Replace the project directory with a placeholder to make it system-independent
    const projectDirectory = path.resolve(__dirname, '..', '..', '..');
    const normalizedPath = val.replace(
      new RegExp(projectDirectory.replace(/\\/g, '\\\\'), 'g'),
      '<PROJECT_DIRECTORY>',
    );
    return normalizedPath;
  },
  test(val) {
    return typeof val === 'string';
  },
});

beforeAll(() => {
  Deluge = new fileSystem();
  dummy_dir = fakeFS();
  console.log(dummy_dir);
});

afterAll(() => {
  Deluge = null;
});

test('filesystem exists', () => {
  const instance = new fileSystem();
  expect(typeof instance == 'object', 'filesystem exists');
});

test('fail on blank', async () => {
  const instance = new fileSystem();
  expect.assertions(1);
  try {
    await instance.init();
  } catch (err) {
    expect(err.message).toBe('path given is not a directory');
  }
});

test('fail on file', async () => {
  const instance = new fileSystem();
  expect.assertions(1);
  try {
    await instance.init(path.join(testfolder, 'blank.txt'), {
      renameToV4: true,
      prettyNames: false,
    });
  } catch (err) {
    expect(err.message).toBe('path given is not a directory');
  }
});

test('fail on non-deluge directory', async () => {
  const instance = new fileSystem();
  expect.assertions(1);
  try {
    await instance.init(path.join(testfolder, 'emptyfolder'), {
      renameToV4: true,
      prettyNames: false,
    });
  } catch (err) {
    expect(err.message).toBe("can't find deluge paths");
  }
});

test('sucessful init', () => {
  return Deluge.init(dummy_dir, {
    renameToV4: true,
    prettyNames: false,
  }).then(() => {
    expect(Deluge.isDelugeDrive).toBe(true);
  });
});

test('check song mappings', () => {
  expect(getMainDelugeInfo(Deluge)).toMatchSnapshot();
});

test('validate each song', () => {
  Deluge.files.songs.forEach(song => {
    expect(song.validate()).toBe(true);
  });
});

test('fail on bad array length', () => {
  expect(() => {
    Deluge.files.songs.forEach(song => {
      const newNames = [];
      song.rewriteInstrumentsAndClipsXMLAttributes(newNames);
    });
  }).toThrow('new names are not the same length as instruments');
});

test('fail on bad array data', () => {
  expect(() => {
    Deluge.files.songs.forEach(song => {
      const newNames = new Array(song.instruments.length).fill({
        rewriteName: false,
        rewriteFolder: 'test folder ',
      });
      song.rewriteInstrumentsAndClipsXMLAttributes(newNames);
    });
  }).toThrow('invalid data in new names array');
});

test('test rewrite each song', () => {
  expect(() => {
    Deluge.files.songs.forEach(song => {
      const newNames = song.instruments.map((_v, index) => ({
        rewriteName: 'test name ' + index,
        rewriteFolder: 'test folder ' + index,
      }));
      song.rewriteInstrumentsAndClipsXMLAttributes(newNames);
    });
  }).not.toThrow();
});

test('find test names in songs', () => {
  Deluge.files.songs.forEach(song => {
    song.instruments.forEach((instrument, index) => {
      expect(instrument.presetName).toBe('test name ' + index);
      expect(instrument.presetFolder).toBe('test folder ' + index);
      expect(instrument.presetSlot).toBeNull;
      expect(instrument.presetSubSlot).toBeNull;
    });
  });
});

test('find test names in clips', () => {
  Deluge.files.songs.forEach(song => {
    song.instruments.forEach((instrument, index) => {
      instrument.clips.forEach(clipId => {
        const clip = song.clips[clipId];
        expect(clip.presetName).toBe('test name ' + index);
        expect(clip.presetFolder).toBe('test folder ' + index);
        expect(clip.presetSlot).toBeNull;
        expect(clip.presetSubSlot).toBeNull;
      });
    });
  });
});

test('no old tags in clips', () => {
  Deluge.files.songs.forEach(song => {
    song.clips.forEach(clip => {
      expect(clip.presetSlot).toBeNull;
      expect(clip.presetSubSlot).toBeNull;
    });
  });
});

/*

TODO:

make a fake mapping directors that includes:
- old - xml song file with dummy presetSlot (factory sound + subSlot ('0' '-1') ('0' '1'), no match + subSlot ('999' '-1') ('999' '1'), ('171','-1') + ('171','1'))



- numsonly - xml song file with dummy nummsonly ('000' + '000A, '999' + '999A', '171' + '171A')



- newsuffix -
- new -
- nameonly -

*/
