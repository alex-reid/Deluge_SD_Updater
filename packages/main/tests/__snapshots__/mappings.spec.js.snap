// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`check song mappings 1`] = `
{
  initialised: true,
  kits: [
    {
      newName: 000 TR-808,
      oldName: KIT000,
      path: KITS,
      presetName: KIT000,
      presetType: KIT,
      prettyName: 000 Kit,
      rewriteName: ,
      rootPath: <PROJECT_DIRECTORY>/fake_filesystem/FAKE_SD_CARD,
    },
    {
      newName: 001 DDD-1,
      oldName: KIT001,
      path: KITS,
      presetName: KIT001,
      presetType: KIT,
      prettyName: 001 Kit,
      rewriteName: ,
      rootPath: <PROJECT_DIRECTORY>/fake_filesystem/FAKE_SD_CARD,
    },
    {
      newName: undefined 3,
      oldName: KIT171B,
      path: KITS,
      presetName: KIT171B,
      presetType: KIT,
      prettyName: 171 Kit B,
      rewriteName: ,
      rootPath: <PROJECT_DIRECTORY>/fake_filesystem/FAKE_SD_CARD,
    },
    {
      newName: ,
      oldName: KIT172,
      path: KITS,
      presetName: KIT172,
      presetType: KIT,
      prettyName: 172 Kit,
      rewriteName: ,
      rootPath: <PROJECT_DIRECTORY>/fake_filesystem/FAKE_SD_CARD,
    },
    {
      newName: undefined 3,
      oldName: KIT272B,
      path: KITS,
      presetName: KIT272B,
      presetType: KIT,
      prettyName: 272 Kit B,
      rewriteName: ,
      rootPath: <PROJECT_DIRECTORY>/fake_filesystem/FAKE_SD_CARD,
    },
  ],
  songs: [
    {
      clipsLength: 6,
      firmwareVersion: [
        0,
        0,
        0,
      ],
      instruments: [
        {
          formatType: old,
          patchName: KIT000,
          presetFolder: null,
          presetName: null,
          presetSlot: 0,
          presetSubSlot: -1,
          rewriteFolder: ,
          rewriteName: ,
          type: kit,
          usedInClips: [
            0,
          ],
        },
        {
          formatType: old,
          patchName: SYNT000,
          presetFolder: null,
          presetName: null,
          presetSlot: 0,
          presetSubSlot: -1,
          rewriteFolder: ,
          rewriteName: ,
          type: sound,
          usedInClips: [
            1,
          ],
        },
        {
          formatType: old,
          patchName: KIT171B,
          presetFolder: null,
          presetName: null,
          presetSlot: 171,
          presetSubSlot: 1,
          rewriteFolder: ,
          rewriteName: ,
          type: kit,
          usedInClips: [
            2,
          ],
        },
        {
          formatType: old,
          patchName: SYNT171B,
          presetFolder: null,
          presetName: null,
          presetSlot: 171,
          presetSubSlot: 1,
          rewriteFolder: ,
          rewriteName: ,
          type: sound,
          usedInClips: [
            3,
          ],
        },
        {
          formatType: old,
          patchName: KIT271,
          presetFolder: null,
          presetName: null,
          presetSlot: 271,
          presetSubSlot: -1,
          rewriteFolder: ,
          rewriteName: ,
          type: kit,
          usedInClips: [
            4,
          ],
        },
        {
          formatType: old,
          patchName: SYNT271,
          presetFolder: null,
          presetName: null,
          presetSlot: 271,
          presetSubSlot: -1,
          rewriteFolder: ,
          rewriteName: ,
          type: sound,
          usedInClips: [
            5,
          ],
        },
      ],
      name: V3_OLD,
      path: SONGS,
    },
    {
      clipsLength: 8,
      firmwareVersion: [
        0,
        0,
        0,
      ],
      instruments: [
        {
          formatType: new,
          patchName: ,
          presetFolder: KITS,
          presetName: KIT001,
          presetSlot: null,
          presetSubSlot: null,
          rewriteFolder: ,
          rewriteName: ,
          type: kit,
          usedInClips: [
            0,
          ],
        },
        {
          formatType: new,
          patchName: ,
          presetFolder: SYNTHS,
          presetName: SYNT001,
          presetSlot: null,
          presetSubSlot: null,
          rewriteFolder: ,
          rewriteName: ,
          type: sound,
          usedInClips: [
            1,
          ],
        },
        {
          formatType: new,
          patchName: ,
          presetFolder: KITS,
          presetName: KIT172,
          presetSlot: null,
          presetSubSlot: null,
          rewriteFolder: ,
          rewriteName: ,
          type: kit,
          usedInClips: [
            2,
          ],
        },
        {
          formatType: new,
          patchName: ,
          presetFolder: SYNTHS,
          presetName: SYNT172,
          presetSlot: null,
          presetSubSlot: null,
          rewriteFolder: ,
          rewriteName: ,
          type: sound,
          usedInClips: [
            3,
          ],
        },
        {
          formatType: new,
          patchName: ,
          presetFolder: KITS,
          presetName: KIT173,
          presetSlot: null,
          presetSubSlot: null,
          rewriteFolder: ,
          rewriteName: ,
          type: kit,
          usedInClips: [
            4,
          ],
        },
        {
          formatType: new,
          patchName: ,
          presetFolder: SYNTHS,
          presetName: SYNT173,
          presetSlot: null,
          presetSubSlot: null,
          rewriteFolder: ,
          rewriteName: ,
          type: sound,
          usedInClips: [
            5,
          ],
        },
        {
          formatType: newsuffix,
          patchName: ,
          presetFolder: KITS,
          presetName: KIT272B,
          presetSlot: null,
          presetSubSlot: null,
          rewriteFolder: ,
          rewriteName: ,
          type: kit,
          usedInClips: [
            6,
          ],
        },
        {
          formatType: newsuffix,
          patchName: ,
          presetFolder: SYNTHS,
          presetName: SYNT272B,
          presetSlot: null,
          presetSubSlot: null,
          rewriteFolder: ,
          rewriteName: ,
          type: sound,
          usedInClips: [
            7,
          ],
        },
      ],
      name: V4_OLDNAMES,
      path: SONGS,
    },
  ],
  synths: [
    {
      newName: 000 Rich Saw Bass,
      oldName: SYNT000,
      path: SYNTHS,
      presetName: SYNT000,
      presetType: SYNT,
      prettyName: 000 Synth,
      rewriteName: ,
      rootPath: <PROJECT_DIRECTORY>/fake_filesystem/FAKE_SD_CARD,
    },
    {
      newName: 001 Sync Bass,
      oldName: SYNT001,
      path: SYNTHS,
      presetName: SYNT001,
      presetType: SYNT,
      prettyName: 001 Synth,
      rewriteName: ,
      rootPath: <PROJECT_DIRECTORY>/fake_filesystem/FAKE_SD_CARD,
    },
    {
      newName: undefined 3,
      oldName: SYNT171B,
      path: SYNTHS,
      presetName: SYNT171B,
      presetType: SYNT,
      prettyName: 171 Synth B,
      rewriteName: ,
      rootPath: <PROJECT_DIRECTORY>/fake_filesystem/FAKE_SD_CARD,
    },
    {
      newName: ,
      oldName: SYNT172,
      path: SYNTHS,
      presetName: SYNT172,
      presetType: SYNT,
      prettyName: 172 Synth,
      rewriteName: ,
      rootPath: <PROJECT_DIRECTORY>/fake_filesystem/FAKE_SD_CARD,
    },
    {
      newName: undefined 3,
      oldName: SYNT272B,
      path: SYNTHS,
      presetName: SYNT272B,
      presetType: SYNT,
      prettyName: 272 Synth B,
      rewriteName: ,
      rootPath: <PROJECT_DIRECTORY>/fake_filesystem/FAKE_SD_CARD,
    },
  ],
}
`;
